OpenCL: helper functions for create/destroy buffers

claudioandre-br opened the pull request:
I started working on #4363 . I was thinking about what should be done and how. 

I'm using raw-sha512 for testing. 
- Inside the new `release_clobj` I call a kind of `clear_all()` function.
```
static void release_clobj()
{
	if (ocl_initialized) {
		CLRELEASEALLMAPBUFFERS(cl_map_list, cl_map_size, CL_HARD_ERROR);
		CLRELEASEALLBUFFERS(cl_buffer_list, cl_buffer_size, CL_HARD_ERROR);
		MEM_FREE(hash_ids);

		HANDLE_CLERROR(clFinish(queue[gpu_id]), "Error releasing memory");
	}
	cl_buffer_size = 0;
	cl_map_size = 0;
	ocl_initialized = 0;
}
```
- The mask mode buffers are different. More regular, I create and destroy buffers when needed, at specific points. So, I create and release the buffers in a traditional way.
```
	buffer_bitmap = CLCREATEBUFFERx2(
	    CL_MEM_WRITE_ONLY, (bitmap_size / 32 + 1) * sizeof(uint32_t), CL_HARD_ERROR);
        [...]
	CLRELEASEBUFFERx(buffer_bitmap, CL_HARD_ERROR);
}
```

Commit:
OpenCL: helper functions for create/destroy buffers

These functions can have memory-leak detection.

ghost commented:
No leaks so far. This is the kind of change I would like to see in `opencl_common.h`.

claudioandre-br commented on src/opencl_common.h:
```diff
@@ -152,6 +152,11 @@ extern int device_info[MAX_GPU_DEVICES];
+#define CL_HARD_ERROR           2
+#define CL_IGNORE_ERROR         3
+#define CL_MAX_BUFFERS         32
+
```
All of these use cases must use `CL_HARD_ERROR`. Remove this.

solardiz replied:
I was thinking the same (before I saw your comment) - we must not ignore this kind of errors, or we'd crash on a dependent call next. So no reason to support ignoring them.

claudioandre-br commented on src/opencl_common.h:
```diff
@@ -274,6 +298,158 @@ void opencl_process_event(void);
+ * Later, it can release all buffers at once
+*/
+#define CLCREATEBUFFERx(_flags, _size, error_level, buffer_list, buffer_size) ({    \
+    cl_mem __buffer;                                                                \
```
I need a better name (`CLCREATEBUFFER_ADD_LIST`)

magnumripper replied:
Keeping it short-ish it could be `CLCREATEBUF_LIST`. It could even be `CLCREATEBUFFERS` vs `CLCREATEBUFFER`...

Perhaps some of these things should be real functions in `opencl_common.c` instead of function-like macros.

claudioandre-br replied:
Which ones?

solardiz replied:
Different perspective: why are any of these macros at all? Shouldn't all of them be functions?

magnumripper replied:
Right!

magnumripper replied:
> Keeping it short-ish it could be `CLCREATEBUF_LIST`. It could even be `CLCREATEBUFFERS` vs `CLCREATEBUFFER`...

I misunderstood your functions/macros yesterday. I thought the one now called `opencl_create_buf_list()` would be able to allocate several/all buffers for a format, in one call. Now I see `_list` refers to "keeping track".

Given that, I'm not sure the name need to reflect that at all (just like `malloc()` keeps lists and whatnot under the hood but it's not called `malloc_list()`). Maybe we should *only* export a function that does keep track, yet just call it `opencl_create_buf()`.

claudioandre-br commented on src/opencl_common.h:
```diff
@@ -274,6 +298,158 @@ void opencl_process_event(void);
+			_out_buffer = NULL;                                                         \
+			_out_map = NULL;                                                            \
+			if (error_level == CL_SOFT_ERROR)                                           \
+				SOFT_CLERROR(__errCreateBuffer, "Error allocating GPU memory");         \
```
Ops, should be `__failure`.

claudioandre-br commented on src/opencl_common.h:
```diff
@@ -274,6 +298,158 @@ void opencl_process_event(void);
+		}                                                                               \
+		if (__failure != CL_SUCCESS) {                                                  \
+			if (error_level == CL_SOFT_ERROR)                                           \
+				SOFT_CLERROR(__errRelease, "Error Releasing buffers");                  \
```
Ops, should be `__failure`.

claudioandre-br commented on src/opencl_common.h:
```diff
@@ -274,6 +298,158 @@ void opencl_process_event(void);
+		}                                                                               \
+		if (__failure != CL_SUCCESS) {                                                  \
+			if (error_level == CL_SOFT_ERROR)                                           \
+				SOFT_CLERROR(__errRelease, "Error Releasing mapped buffers");           \
```
Ops, should be `__failure`.

ghost commented:
Better to see the new macros and an example of use.

solardiz commented on src/opencl_rawsha512_gpl_fmt_plug.c:
```diff
@@ -58,6 +58,11 @@ static cl_mem pass_buffer;      //Plaintext buffer.
+//A list containing all OpenCL buffers created by the format.
+static cl_mem cl_buffer_list[CL_MAX_BUFFERS];
+static void *cl_map_list[CL_MAX_BUFFERS][2];
+static int cl_buffer_size, cl_map_size;
```
Should we possibly provide such helpers that we'd only need to define one variable here (of a custom struct type) and the helpers would handle allocation and freeing of the buffer list dynamically? Perhaps it can be a couple of linked lists managed by the helpers, with list tails in the struct.

BTW, if we end up freeing in the reverse order, that can actually be friendly towards the underlying memory allocators.

magnumripper replied:
Something like that is boiling in my head too: A single define for all buffers that depend on GWS. Then a single call to allocate all of them and a single call to destroy them.

Perhaps even a single define for all buffers, where some depend on GWS and some don't? We need to keep scope in mind though.

What is the scope? I think it's "minimize (as in share) code that all formats need", but also - and this is important: "make things easy to read and follow".

ghost commented:
I refactored and pushed the sha512 source code (make it easier to maintain).

Now I'm moving the code from macros to functions.

ghost commented:
Please, comment the "C"ism;
* I can't think about more automation. Or, I don't need it. I'm happy; that is my use case;
* formats create only a few buffers. This first version does not have linked list and dynamic allocation;

IMO, the change is invasive. Macros would be a smoother transition.

ghost commented:
Now:
* using a linked list;
* one createBufferOpenCL API;
```
salt = opencl_create_buf(RO, salt_size, keep_track=1);
mask = opencl_create_buf(RO, mask_size, keep_track=0);
```
Working on (bold):
> we could have it as opencl_create_buf(&buf, flags, size). This makes it easier to optionally verify it was NULL before the alloc and we can keep track of buf in our list so upon **calling opencl_release_buffers() we'd set it to NULL even though it wasn't given as an argument**.

ghost commented:
I will not add new resources to the PR, only refactor/rename/organize code. I tested for leaks and in a real cracking session.

* it has buffers, pinned buffers and kernels creation. Uses linked list (not yet for `map()`).

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2881,269 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+			if (list->tail == current)
+				list->tail = prev;
+
+			*(current->variable) = NULL; //Is it really safe?
```
This is no worse than what `MEM_FREE(buf)` do on host side

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2881,269 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	cl_int __errCreateBuffer;
+
+	if (*(_buffer))
+		opencl_release_buf(_buffer);
```
We could regard this as an error instead. Or, we could make that depend on `-DDEBUG`

claudioandre-br replied:
Done

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2881,269 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+		_size, NULL, &__errCreateBuffer);
+	*(_out_map) = clEnqueueMapBuffer(queue[gpu_id], *(_out_buffer),
+		_map_block, _map_flags, _map_offset, _map_size, 0, NULL, NULL,
+		&__errMapBuffer);
```
Maybe I'm stupid now but isn't (for our use) always `_size` same as `_map_size`? If so we could drop the latter argument from the function.

Also, I **think** we (so far) never use `CL_MAP_READ` so perhaps we should hard-code that (perhaps even to `CL_​MAP_​WRITE_​INVALIDATE_​REGION`?) and drop the `_map_flags` argument.

magnumripper replied:
All current code has `_map_block` as `CL_TRUE` so we can optimize that away as well

magnumripper replied:
All current code has `_map_offset` as `0` so we can optimize that away as well

claudioandre-br replied:
Done.

Commit:
X/raw-sha512-opencl: refactor and use shared mem code

Start using shared macros and functions to create and release OpenCL
memory buffers and kernels.

claudioandre-br commented on src/opencl_common.c:
```diff
@@ -88,6 +88,54 @@ extern int autotune_get_next_gws_size(size_t num, int step, int startup,
 
+/*
+ * List entry, allocated as (sizeof(struct kernel_opencl) + sizeof(data)).
+ */
```
To avoid duplication, I can use pseudo-generics. But, the source code becomes hard to follow.

claudioandre-br commented on src/opencl_common.c:
```diff
@@ -2829,6 +2881,273 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
 	return gws * v_width;
 }
 
+static void buf_list_init(struct opencl_buffers **list) {
```
Idem. To avoid duplication, I can use pseudo-generics. But, the source code becomes hard to follow.

magnumripper replied:
Very short and concise functions, probably better to keep them as-is for readability.

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2881,273 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	*(out_buffer) = clCreateBuffer(context[gpu_id], buffer_flags,
+		size, NULL, &err_create_buffer);
+	*(out_map) = clEnqueueMapBuffer(queue[gpu_id], *(out_buffer),
+		CL_TRUE, CL_MAP_WRITE | CL_MAP_WRITE_INVALIDATE_REGION, 0, size, 0, NULL, NULL,
```
If I got it right, the `CL_MAP_WRITE` is superfluous here but I'm not 100% sure

claudioandre-br replied:
Reading again, they are mutually exclusive. So, yes.

claudioandre-br replied:
BTW: it is 1.2, is everyone okay with this?

magnumripper replied:
We have this in `opencl_common.h`
```c
#define CL_TARGET_OPENCL_VERSION 120
```

magnumripper commented on src/opencl_rawsha512_gpl_fmt_plug.c:
```diff
@@ -162,33 +196,49 @@ static uint64_t *crypt_one_x(int index) {
-	MEM_FREE(saved_bitmap);
+	release_bitmap_buffers();
+
+	bitmap = (uint32_t *)
+		mem_alloc((bitmap_size / 32 + 1) * sizeof(uint32_t));
+	opencl_create_buf(&dev_bitmap,
+	    CL_MEM_WRITE_ONLY, (bitmap_size / 32 + 1) * sizeof(uint32_t), CL_DONT_KEEP_TRACK);
```
Shouldn't most our functions (or new alternatives to them) cover both?

Something like:
```c
opencl_create_bufs(&bitmap, &dev_bitmap, RO, (bitmap_size / 32 + 1) * sizeof(uint32_t), CL_NO_TRACK);
```
This would cover lines 225-228: Do a malloc for `bitmap` and an OpenCL create for `dev_bitmap` in a single call. With future means for great debugging assertions.

claudioandre-br replied:
Done.

claudioandre-br replied:
We have a problem here. At least one: salt. We usually don't `malloc()` salt.

magnumripper replied:
So we need alternative functions: Ones that only allocates a device buffer, and ones that allocate a host buffer and a device buffer with same size.

I guess the `_map` functions can always do both though? Or maybe they already do.

magnumripper replied:
It may seem overkill but in my head it will simplify 99% of the formats so I think we should really consider these things. That being said, I'm just brainstorming.

claudioandre-br replied:
If you think that `MEM_FREE()` is going to happen automagically, it looks good.

magnumripper replied:
Absolutely. Perhaps we can call this `opencl_create_buf_pair`.  On call to `opencl_release_buf_pairs()` - or whatever we'll call this - we'd release both.

magnumripper replied:
Now, we could make these `_pair` functions yet another set of similar functions, or we could do half of them with macros. I'm still merely brainstorming. It could go something like this (just from head):
```c
cl_mem opencl_create_buf_pair(void **host_buf cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track)
{
	cl_int err_create_buffer;

	if (*(buffer_obj))
#ifdef DEBUG
		error_msg("** %s:%d Memory leak (device buffer)\n", __FILE__, __LINE__);
#else
		opencl_release_buf(buffer_obj);
#endif

	if (*(host_buf) && **(host_buf))
#ifdef DEBUG
		error_msg("** %s:%d Memory leak (host buffer)\n", __FILE__, __LINE__);
#else
		MEM_FREE(**host_buf);
#endif

	*(buffer_obj) = clCreateBuffer(context[gpu_id], flags, size, NULL, &err_create_buffer);

	if (err_create_buffer != CL_SUCCESS) {
		*(buffer_obj) = NULL;
		HANDLE_CLERROR(err_create_buffer, "Error allocating GPU memory");
	} else
		if (keep_track)
			buf_list_add(cl_buffer_list, buffer_obj);

	if (*host_buf)
		*(host_buf) = mem_alloc(size);

	return *(buffer_obj); /* or should be drop this and just be void? Or return an error code? The buffer was in the parameters */
 }

/* Version for device-side only */
#define opencl_create_buf(buffer_obj, flags, size, keep_track) \
	opencl_create_buf_pair(NULL, buffer_obj, flags, size, keep_track)
```

magnumripper replied:
BTW since `opencl_create_buf_*()` are no longer macros, we lose the value of `__FILE__, __LINE__`. We could fix that by wrapping them with master macros much like `error_msg()` et al.

claudioandre-br replied:
If it prints `opencl_common.c line 3007` it is ok to me. The same failure/message can happen in some places, the message will be useful. Anyone can say something about macroization?

magnumripper replied:
It's trivial really.

If the real function is
```c
cl_mem opencl_create_buf_pair(void **host_buf, cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track)
```
we modify it to be
```c
cl_mem real_opencl_create_buf_pair(void **host_buf, cl_mem *buffer_obj, cl_mem_flags flags, size_t size, int keep_track, char *file, int line)
```
and use a macro for calling it
```c
#define opencl_create_buf_pair(host_buf, buffer_obj, flags, size, keep_track) \
    real_opencl_create_buf_pair(host_buf, buffer_obj, flags, size, keep_track, __FILE__, __LINE__)
```

ghost commented:
Sorry, for the spam. BTW: this is ready for review.

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2882,291 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
 	return gws * v_width;
 }
 
+static void buf_list_init(struct opencl_buffers **list) {
```
(Many places)

Our code style (wherever it's stated, if it is at all) says this (for functions only)
```diff
- static void buf_list_init(struct opencl_buffers **list) {
+ static void buf_list_init(struct opencl_buffers **list)
+ {
```

magnumripper commented on src/opencl_common.c:
```diff
@@ -2829,6 +2882,291 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	if (err_create_buffer != CL_SUCCESS)
+		failure = err_create_buffer;
+	if (err_map_buffer != CL_SUCCESS)
+		failure = err_map_buffer;
```
If `clCreateBuffer()` fails with error, shouldn't we catch it and bail *before* going on with `clEnqueueMapBuffer()`? Maybe I overlook something but I can't see the point of keeping separate error variables and chug along.
Also, `ret_code` is already globally declared in opencl_common.c so why not use that one:
```c
	(...)
	*(out_buffer) = clCreateBuffer(context[gpu_id], buffer_flags,
		size, NULL, &ret_code);
	HANDLE_CLERROR(ret_code, "Error allocating GPU memory");

	*(out_map) = clEnqueueMapBuffer(queue[gpu_id], *(out_buffer),
		CL_TRUE, CL_MAP_WRITE_INVALIDATE_REGION, 0, size, 0, NULL, NULL,
		&ret_code);
	HANDLE_CLERROR(ret_code, "Error allocating GPU memory");

	cl_map_list[cl_map_size][0] = *(out_buffer);
	cl_map_list[cl_map_size++][1] = *(out_map);
```
The above will also let the `HANDLE_CLERROR` give (more) correct line numbers for failure, instead of always stating it was line 3059 (BTW we should make them refer to the caller's file:line instead, perhaps I should add that to this PR? Fairly trivial.)

magnumripper commented on src/opencl_rawsha512_gpl_fmt_plug.c:
```diff
@@ -197,145 +246,56 @@ static void create_clobj(size_t gws, struct fmt_main *self)
 		mask_gws = gws;
 	}
+	// create pinned buffers
+	opencl_create_map(
+		CL_MEM_READ_ONLY | CL_MEM_ALLOC_HOST_PTR, BUFFER_SIZE * gws,
+		&pinned_plaintext,     /* buffer created on device */
+		(void *) &plaintext);  /* mapped buffer created on host */
```
(Many places)

No need to cast to `void*`, that's sort of the idea with using it as function parameter!

claudioandre-br replied:
```
opencl_rawsha512_gpl_fmt_plug.c:255: warning: passing argument 4 of ‘opencl_create_map’ from incompatible pointer type
opencl_common.h:306: note: expected ‘void **’ but argument is of type ‘uint32_t **’

```

magnumripper replied:
Oh I didn't know that, pardon me. Apparently it's not allowed with a double pointer. I can't fathom why.

solardiz commented:
Actually switching to these helper functions across all OpenCL formats is a fairly invasive change. Maybe we should first work on fixing the known bugs (most notably, #4044), make a release, and only then merge these changes.

magnumripper commented:
> Actually switching to these helper functions across all OpenCL formats is a fairly invasive change. Maybe we should first work on fixing the known bugs (most notably, #4044), make a release, and only then merge these changes.

That's not a reason for waiting with merging this. It opens up for switching the rest of the formats whenever someone gets the energy, but it doesn't mandate it.

solardiz commented:
OK, I don't mind merging this PR (shared code + one example).

I notice the linked list implementation is copy-paste-edit from our generic `list[.ch]` code - even the comments are copied, but are no longer right for the modified structs/code. I think instead of copy-paste-edit, we should literally use our shared generic `list[.ch]` code - it's meant to provide an opaque linked list implementation, where we only need to define a custom struct for the `data` field. We can use typecasts to pointers to our custom struct, which is OK per C aliasing rules as long as that's the only way we access the data. That's how we already use this generic/opaque implementation in many places - grep the source tree for "struct list" to find those.

solardiz commented:
> That's how we already use this generic/opaque implementation in many places

Upon a closer look, all of our current uses of that implementation are for lists of strings, not of arbitrary binary structs, and the comments say so and the `list_*` functions actually assume C strings. So perhaps the copy-paste-edit is right, but we need to revise the comments accordingly.

Longer-term and separately, we could add a `list_add_binary` function, etc. and start to reuse the shared implementation as truly opaque.

solardiz commented on src/opencl_common.c:
```diff
@@ -88,6 +88,55 @@ extern int autotune_get_next_gws_size(size_t num, int step, int startup,
+
+	cl_kernel data;
+	cl_kernel *variable;
+};
```
Why does the comment say ", allocated as (sizeof(struct kernel_opencl) + sizeof(data))" when the struct itself has a fully-sized field `data`? I think that comment should be dropped, and the allocation simplified.

solardiz commented on src/opencl_common.c:
```diff
@@ -88,6 +88,55 @@ extern int autotune_get_next_gws_size(size_t num, int step, int startup,
+	cl_mem data;
+	cl_mem *variable;
+	void **host_ptr;
+};
```
Ditto about `data`.

solardiz commented on src/opencl_common.c:
```diff
@@ -88,6 +88,55 @@ extern int autotune_get_next_gws_size(size_t num, int step, int startup,
+	struct buffer_opencl *head, *tail;
+
+	int count;
+};
```
I think it'd be convenient for us to start all of these struct names with some prefix indicating they're ours and come from this source file. Like `john_opencl_kernel`, `john_opencl_kernels`, `john_opencl_buffer`, `john_opencl_buffers`. This will become especially important, and will save us the need to rename them later, if we ever have to make them visible to other source files.

magnumripper replied:
I've been pushing the use of `opencl_` to indicate this source file. Khronos OpenCL stuff start with `cl_`.

solardiz replied:
OK, if `opencl_` is our convention so far I don't insist on also having `john_`.

solardiz commented on src/opencl_common.c:
```diff
@@ -2829,6 +2883,296 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+
+	entry = mem_alloc(sizeof(struct buffer_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
```
I think the ` + sizeof(data)` merely wastes memory - we never actually write beyond the size of the struct here, right?

Also, why do we store both `*data` and `data`? In other words, we store the same thing by value and by reference. Why?

Finally, assuming we need it, lets drop the braces in `*(data)`.

claudioandre-br replied:
> Also, why do we store both *data and data? In other words, we store the same thing by value and by reference. Why?

Think about 

1. the OpenCL buffer, the buffer itself, and;
2. the place where the format that made the call stores the information. 

What is of interest to the shared code is the OpenCL buffer, the reference is only for cleaning (assign NULL).

solardiz commented on src/opencl_common.c:
```diff
@@ -2829,6 +2883,296 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+
+	entry = mem_alloc(sizeof(struct kernel_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
```
Same comments as for `buf_list_add`.

solardiz commented on src/opencl_common.c:
```diff
@@ -2829,6 +2883,296 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	struct buffer_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
```
Why do we choose to find the thing to remove by value?

claudioandre-br replied:
To allow a `release_buf(pointer)` in the future.

solardiz replied:
OK. We might or might not need that. But anyway, why search by value?

solardiz commented on src/opencl_common.c:
```diff
@@ -2829,6 +2883,296 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	struct kernel_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
```
Why do we choose to find the thing to remove by value?

solardiz reviewed (changes requested):
At least the weird overallocations need to be fixed.

solardiz commented:
Why do we introduce arbitrary list remove functions for the buffers and the kernels? Is this for some kind of special cases? I think normally we'll just be deallocating things in reverse order, so we might need more like a "free up to that entry" function (and if passed NULL, then free all).

ghost commented:
> Why do we introduce arbitrary list remove functions for the buffers and the kernels? Is this for some kind of special cases? I think normally we'll just be deallocating things in reverse order, so we might need more like a "free up to that entry" function (and if passed NULL, then free all).

IMO, it's the opposite. Formats create and destroy en bloc (usually). Other opinions are welcome.
